File: coding/src/com/algos/core/li09_linkedlist/revised/FlatteningOfALinkedList.java
package  com.algos.core.li09_linkedlist.revised;

import com.algos.core.models.ListNode;

public class FlatteningOfALinkedList {
    public static void main(String[] args) {
        FlatteningOfALinkedList obj = new FlatteningOfALinkedList();
        ListNode head = new ListNode(5);
        head.bottom = new ListNode(7);
        head.bottom.bottom = new ListNode(8);
        head.bottom.bottom.bottom = new ListNode(30);

        head.next = new ListNode(10);
        head.next.bottom = new ListNode(20);

        head.next.next = new ListNode(19);
        head.next.next.bottom = new ListNode(22);
        head.next.next.bottom.bottom = new ListNode(50);

        head.next.next.next = new ListNode(28);
        head.next.next.next.bottom = new ListNode(35);
        head.next.next.next.bottom.bottom = new ListNode(40);

        head = obj.flatten(head);
        while (head != null) {
            System.out.print(head.val + " ");
            head = head.bottom;
        }

    }

    public ListNode flatten(ListNode root) {
        if (root == null || root.next == null)
            return root;

        ListNode mergedList = flatten(root.next);
        return mergeTwoLists(root, mergedList);
    }

    private ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode curr = dummy;

        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                curr.bottom = l1;
                curr = curr.bottom;
                l1 = l1.bottom;
            } else {
                curr.bottom = l2;
                curr = curr.bottom;
                l2 = l2.bottom;
            }
        }

        if (l1 != null)
            curr.bottom = l1;
        else
            curr.bottom = l2;
        return dummy.bottom;

    }
}

File: dsa/concepts/09_linkedlist/python/flattening_of_alinked_list.py
"""
Idea:
Each node has two pointers:
- next  → points to the next list
- bottom → points to a sorted linked list

Flatten the list by:
1. Recursively flattening the right (next) list
2. Merging the current list with the flattened list
"""

class ListNode:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.bottom = None


def flatten(root: ListNode) -> ListNode:
    if root is None or root.next is None:
        return root

    # Flatten the list on the right
    merged_list = flatten(root.next)

    # Merge current list with the flattened list
    return merge_two_lists(root, merged_list)


def merge_two_lists(l1: ListNode, l2: ListNode) -> ListNode:
    dummy = ListNode(0)
    curr = dummy

    while l1 and l2:
        if l1.val < l2.val:
            curr.bottom = l1
            l1 = l1.bottom
        else:
            curr.bottom = l2
            l2 = l2.bottom
        curr = curr.bottom

    # Attach remaining nodes
    curr.bottom = l1 if l1 else l2

    return dummy.bottom


# Example usage
if __name__ == "__main__":
    head = ListNode(5)
    head.bottom = ListNode(7)
    head.bottom.bottom = ListNode(8)
    head.bottom.bottom.bottom = ListNode(30)

    head.next = ListNode(10)
    head.next.bottom = ListNode(20)

    head.next.next = ListNode(19)
    head.next.next.bottom = ListNode(22)
    head.next.next.bottom.bottom = ListNode(50)

    head.next.next.next = ListNode(28)
    head.next.next.next.bottom = ListNode(35)
    head.next.next.next.bottom.bottom = ListNode(40)

    head = flatten(head)

    # Print flattened list using bottom pointer
    while head:
        print(head.val, end=" ")
        head = head.bottom
